### ISSUE TITLE
How do you document `{ json:api }`? (jsonapi.org)

### ISSUE THREAD

**Issue by NullVoxPopuli** [ad-on-is/adonis-autoswagger#144]
I have custom serializers for my data, following jsonapi.org (the only real REST format worth using).

but I'm not sure how to get the autoswagger generation to use my serializers.

how would I do this?

thanks!

**Comment by ad-on-is** [comment]
It's in the readme

**Comment by NullVoxPopuli** [comment]
there is no information on https://jsonapi.org/ -formatted APIs in the README.

or, rather, anywhere on the internet really. It's super weird. It feels like everyone is just handrolling this

**Comment by ad-on-is** [comment]
I'm not familiar eith jsonapi ata all, I had a quick look at it, and I assume it's just a "descriptor" in what format to return responses? 

Please correct me if I'm wrong.

Since you said you created your serializer according to jsonapi, I thought you created a ValidatorClass schema baded on it

**Comment by NullVoxPopuli** [comment]
> and I assume it's just a "descriptor" in what format to return responses?

sorta, ya -- it's very prescriptive about relationships, query params, bulk operations, etc
A GraphQL of sorts, but without throwing away all the features of REST. lol


> Since you said you created your serializer according to jsonapi, I thought you created a ValidatorClass schema baded on it

I'm pretty new to adonis, and am probably doing a million things wrong, actually.

I don't have serializers, or validator classes. :grimacing: 
I suppose I should look into those things

**Comment by ad-on-is** [comment]
> I don't have serializers, or validator classes

Well, there's the issue, this whole project revolves around adonisJs 😅

**Comment by NullVoxPopuli** [comment]
ya, tbf, in adonis' docs, I couldn't find much on jsonapi.org-formatted things there either :thinking: 

maybe I need to make a library

**Comment by ad-on-is** [comment]
That's because jsonapi is just a descriptor, not more  ot less... it's the same as if I tell you right now to return only this format
 
```js
{data: [{name: "Foo"}], meta:{pages:10, current: 1}}
```

why would adonisJs care about what I say and mention it, or even add it into their code? they give you the tools to implement it yourself, if you wish to do so.

**Comment by NullVoxPopuli** [comment]
I'm probably still missing something -- but it's not just a descriptor, it requires you efficiently integrate with the ORM to prevent run-away / N+1 query issues.

Like, jsonapi states that queries coming in with `GET /books?include=books.author,books.publisher,books.characters` includes the author, publisher, and all characters for every book record returned -- so -- at least to me, it feels like there is more integration than just serializing models for openapi docs. The relationship graph being known is helpful

> why would adonisJs care about what I say and mention it, or even add it into their code? 

oh idk, that's up to them -- if there are docs how to configure something like this, I'd be all over them.

> they give you the tools to implement it yourself, if you wish to do so.

I do!

**Comment by ad-on-is** [comment]
> it requires you efficiently integrate with the ORM to prevent run-away / N+1 query issues

Exectly what I said.. it requires YOU to do everything in your app to please their recommendations, which doesn't necessarily mean everyone agrees with them, right?

Since ORMs are just wrappers around SQL queries, you can always write raw SQL, if you hit the limits to please the jsonapi requirements.

Today it's jsonapi, tomorrow it could be "theBestApiSpecIntheworld" claiming totally different things than jsonapi .. it's up to the developer whether they want to do it this way or not, and no need for core libraries to add it for convenience purposes.

**Comment by NullVoxPopuli** [comment]
> it requires YOU to do everything in your app to please their recommendations, which doesn't necessarily mean everyone agrees with them, right?

no, it's a spec. if you use jsonapi.org, you must implement certain things so that clients who know the spec can have their expectations met.

Rails' jsonapi libraries do a really good job of this with adding serializers and integration with ActiveRecord.

(very minimal config is needed, usually)

the Rails jsonapi support _also_ isn't core.

**Comment by ad-on-is** [comment]
> the Rails jsonapi support also isn't core.

That's also exactly what I just said 😅 So someone went out their way and created a wrapper for Rails to make it more convenient to return things according to jsonapi spec, right?

**Comment by NullVoxPopuli** [comment]
Ye, I'm just saying I'm willing to also do that once i learn enough about how all this stuff plugs together
